docs/settings.md
Settings reference
==================

In Misago there are two types of settings: static settings that are initialized on Misago startup and don't change afterwards, and dynamic settings that are stored in database and can be changed from admin control panel.


Static settings
---------------

Static settings values are accessible throught the attributes of `misago.conf.settings` object:

- `debug: bool` - Switch controlling if `debug` mode is enabled.
- `test: bool` - Switch specifying if application is running within test runner (pytest).
- `test_database_name: Optional[str]` - If specified, makes tests run against different database.
- `database_url: str` - Database connection URL as used by [`Databases` library](https://www.encode.io/databases/database_queries/).
- `cache_url: str` - Cache connection URL as used by [`async-caches` library](https://rafalp.github.io/async-caches/backends/).
- `static_root: str` - Absolute path to directory from which static files are served by the HTTP server.
- `media_root: str` - Absolute path to directory from which uploaded files are served by the HTTP server.
- `avatar_sizes: List[int]` - List of user avatar sizes supported on site. Env variable should be comma separated list: `40, 100, 400`.
- `enabled_plugins: Optional[str]` - Absolute path to text file with a list of enabled plugins.


Dynamic settings
----------------

- `avatar_upload_max_size: int` - Limit of uploaded image file size, in kilobytes.
- `bulk_action_limit: int` - Maximum allowed number of items in bulk action (close threads, delete users).
- `enable_site_wizard: bool` - Controls if site wizard UI and `setupSite` graphQL mutations are available.
- `forum_index_header: str` - Header text on forum index.
- `forum_index_threads: bool` - If `True`, display threads list on forum index. Categories otherwise.
- `forum_index_title: str` - Page title on forum index.
- `forum_name: str` - Forum name.
- `jwt_exp: int` - Time for which JWT tokens stay valid, in seconds.
- `jwt_secret: str` - JWT secret. Should be at least 64 characters long to be secure.
- `password_min_length: int` - Min. allowed password length.
- `post_min_length: int` - Min. required length of post body.
- `posts_per_page: int` - Number of posts displayed per page on thread view.
- `posts_per_page_orphans: int` - Max. number of orphaned posts that should rolled up to previous page.
- `thread_title_min_length: int` - Min. required length of thread title.
- `thread_title_max_length: int` - Max. allowed length of thread title.
- `threads_per_page: int` - Max number of threads displayed on single page of threads list.
- `username_max_length: int` - Min. allowed username length.
- `username_min_length: int` - Max. allowed username length.

docs/ref/Settings.md
Settings
========


## Core Settings

Those settings are set in `settings.py` file and available via `django.conf.settings` or `misago.conf.settings`


### MISAGO_403_IMAGE

Url (relative to `STATIC_URL`) to file that should be served if user has no permission to see requested attachment.


### MISAGO_404_IMAGE

Url (relative to `STATIC_URL`) to file that should be served if user has requested nonexistant attachment.


### MISAGO_ACL_EXTENSIONS

List of Misago ACL framework extensions.


### MISAGO_ADMIN_NAMESPACES

Link namespaces that are administrator-only areas that require additional security from Misago. Users will have to re-authenticate themselves to access those namespaces, even if they are already signed in your frontend. In addition they will be requested to reauthenticated if they were inactive in those namespaces for certain time.

Defautly `misago:admin` and `admin` namespaces are specified, putting both Misago and Django default admin interfaces under extended security mechanics.


### MISAGO_ADMIN_PATH

Path prefix for Misago administration backend. Defautly "admincp", but you may set it to empty string if you with to disable your forum administration backend.


### MISAGO_ADMIN_SESSION_EXPIRATION

Maximum allowed lenght of inactivity period between two requests to admin namespaces. If its exceeded, user will be asked to sign in again to admin backed before being allowed to continue activities.


### MISAGO_ATTACHMENT_IMAGE_SIZE_LIMIT

Max dimensions (width and height) of user-uploaded images embedded in posts. If uploaded image is greater than dimensions specified in this settings, Misago will generate thumbnail for it.


##### Note

Because user-uploaded GIF's may be smaller than dimensions specified, but still be considerably heavy due to animation, Misago always generates thumbnails for user-uploaded GIFS, stripping the animations from them.


### MISAGO_ATTACHMENT_ORPHANED_EXPIRE

How old (in minutes) should attachments unassociated with any be before they'll automatically deleted by `clearattachments` task.


### MISAGO_ATTACHMENT_SECRET_LENGTH

Length of attachment's secret (filenames and url token). The longer, the harder it is to bruteforce, but too long may conflict with your uploaded files storage limits (eg. filesystem path length limits).


##### Warning

In order for Misago to support clustered deployments or CDN's (like Amazon's S3), its unable to validate user's permission to see the attachment at its source. Instead it has to rely on exessively long and hard to guess urls to attachments and assumption that your users will not "leak" source urls to attachments further.

Generaly, neither you nor your users should use forums to exchange files containing valuable data, but if you do, you should make sure to secure it additionaly via other means like password-protected archives or file encryption solutions.


### MISAGO_AVATAR_GALLERY

Path to directory containing avatar galleries. Those galleries can be loaded by running `loadavatargallery` command.

Feel free to remove existing galleries or add your own.

If you create gallery named `__default__` and set avatar gallery as default user avatar, Misago will select new users avatars from it while keeping this gallery hidden from existing users.


### MISAGO_AVATARS_SIZES

Misago uses avatar cache that prescales avatars to requested sizes. Enter here sizes to which those should be optimized.


##### Warning

It's impossible to regenerate user avatars store for existing avatars. Misago comes with sane defaults for avatar sizes, with min. height for user avatar being 400 pixels square, and steps of 200, 150, 100, 64, 50 and 30px. However if you need larger avatar or different pregenerated dimensions, changing those will require you to manually remove `avatars` directory from your media storage as well as running `misago.users.avatars.set_default_avatar` function against every user registered.


### MISAGO_BLANK_AVATAR

This path to image file that Misago should use as blank avatar.


### MISAGO_COMPACT_DATE_FORMAT_DAY_MONTH

Date format used by Misago `compact_date` filter for dates in this year.

Expects standard Django date format, documented [here](https://docs.djangoproject.com/en/dev/ref/templates/builtins/#date)


### MISAGO_COMPACT_DATE_FORMAT_DAY_MONTH_YEAR

Date format used by Misago `compact_date` filter for dates in past years.

Expects standard Django date format, documented [here](https://docs.djangoproject.com/en/dev/ref/templates/builtins/#date)


### MISAGO_DIALY_POST_LIMIT

Dialy limit of posts that may be posted from single account. Fail-safe for situations when forum is flooded by spam bot. Change to 0 to lift this restriction.


### MISAGO_DYNAMIC_AVATAR_DRAWER

Function used to create unique avatar for this user. Allows for customization of algorithm used to generate those.


### MISAGO_EVENTS_PER_PAGE

Misago reads events to display in separate database query to avoid situation when thread with large number of eg. moderator actions displays pages consisting exclusively of events. Using this setting you may specify upper limit of events displayed on thread's single page. This setting is intented as fail safe, both to save threads from excessively long lists of events your users will have to scroll trough, as well as to keep memory usage within limts.

In case of more events than specified being found, oldest events will be truncated.


### MISAGO_HOURLY_POST_LIMIT

Hourly limit of posts that may be posted from single account. Fail-safe for situations when forum is flooded by spam bot. Change to 0 to lift this restriction.


### MISAGO_LOGIN_API_URL
URL to API endpoint used to authenticate sign-in credentials. Musn't contain api prefix or wrapping slashes. Defaults to 'auth/login'.


### MISAGO_MARKUP_EXTENSIONS

List of python modules extending Misago markup.


### MISAGO_NEW_REGISTRATIONS_VALIDATORS

List of functions to be called when somebody attempts to register on forums using registration form.


### MISAGO_NOTIFICATIONS_MAX_AGE

Max age, in days, of notifications stored in database. Notifications older than this will be delted.


### MISAGO_POST_ATTACHMENTS_LIMIT

Limit of attachments that may be uploaded in single post. Lower limits may hamper image-heavy forums, but help keep memory usage by posting process. 


### MISAGO_POSTING_MIDDLEWARES

List of middleware classes participating in posting process.


### MISAGO_POSTS_PER_PAGE

Controls number of posts displayed on thread page. Greater numbers can increase number of objects loaded into memory and thus depending on features enabled greatly increase memory usage.


### MISAGO_POSTS_TAIL

Defines minimal number of posts for thread's last page. If number of posts on last page is smaller or equal to one specified in this setting, last page will be appended to previous page instead.


### MISAGO_RANKING_LENGTH

Some lists act as rankings, displaying users in order of certain scoring criteria, like number of posts or likes received.
This setting controls maximum age in days of items that should count to ranking.


### MISAGO_RANKING_SIZE

Maximum number of items on ranking page.


### MISAGO_READTRACKER_CUTOFF

Controls amount of data used by readtracking system. All content older than number of days specified in this setting is considered old and read, even if opposite is true. Active forums can try lowering this value while less active ones may wish to increase it instead.


### MISAGO_SEARCH_CONFIG

PostgreSQL text search configuration to use in searches. Defaults to "simple", for list of installed configurations run "\dF" in "psql".

Standard configs as of PostgreSQL 9.5 are: `dutch`, `english`, `finnish`, `french`, `german`, `hungarian`, `italian`, `norwegian`, `portuguese`, `romanian`, `russian`, `simple`, `spanish`, `swedish`, `turkish`.

Example on adding custom language can be found [here](https://github.com/lemonskyjwt/plpstgrssearch).


##### Note

Items in Misago are usually indexed in search engine on save or update. If you change search configuration, you'll need to rebuild search for past posts to get reindexed using new configuration. Misago comes with `rebuildpostssearch` tool for this purpose.


### MISAGO_STOP_FORUM_SPAM_MIN_CONFIDENCE

Minimum confidence returned by [Stop Forum Spam](http://www.stopforumspam.com/) for Misago to reject new registration and block IP address for 1 day.


### MISAGO_THREADS_ON_INDEX

Change this setting to `False` to display categories list instead of threads list on board index.


### MISAGO_THREADS_PER_PAGE

Controls number of threads displayed on page. Greater numbers can increase number of objects loaded into memory and thus depending on features enabled greatly increase memory usage.


### MISAGO_THREADS_TAIL

Defines minimal number of threads for lists last page. If number of threads on last page is smaller or equal to one specified in this setting, last page will be appended to previous page instead.


### MISAGO_THREAD_TYPES

List of clasess defining thread types.


### MISAGO_USE_STOP_FORUM_SPAM

This settings allows you to decide wheter of not [Stop Forum Spam](http://www.stopforumspam.com/)database should be used to validate IPs and emails during new users registrations.


### MISAGO_USERS_PER_PAGE

Controls pagination of users lists.


# Database Settings

Those settings are stored in database and are available via `misago.conf.settings`. To change those settings you need to use admin control panel


### account_activation

Preffered way in which new user accounts are activated. Can be either of those:

* `none` - no activation required.
* `user` - new user has to click link in activation e-mail.
* `admin` - board administrator has to activate new accounts manually.
* `block` - turn new registrations off.


### allow_custom_avatars

Controls if users may set avatars from outside forums.


### avatar_upload_limit

Max allowed size of uploaded avatars in kilobytes.


### default_avatar

Default avatar assigned to new accounts. Can be either `initials` for randomly generated pic with initials, `gravatar` or `gallery` which will make Misago pick random avatar from gallery instead.


### default_timezone

Default timezone used by guests and newly registered users that haven't changed their timezone prefferences.


### forum_branding_display

Controls branding's visibility in forum navbar.


### forum_branding_text

Allows you to include text besides brand logo on your forum.


### forum_name

Forum name, displayed in titles of pages.


### forum_index_meta_description

Forum index Meta Description used as value meta description attribute on forum index.


### forum_index_title

Forum index title. Can be empty string if not set, in which case `forum_name` should be used instead.


### password_complexity

Complexity requirements for new user passwords. It's value is list of strings representing following requirements:

* `case` - mixed case.
* `alphanumerics` - both digits and letters.
* `special` - special characters.


### password_length_min

Minimal required length of new user passwords.


### post_length_max

Maximal allowed post content length.


### post_length_min

Minimal allowed post content length.


### signature_length_max

Maximal allowed length of users signatures.


### subscribe_reply

Default value for automatic subscription to replied threads prefference for new user accounts. Its value represents one of those settings:

* **no** - don't watch.
* **watch** - put on watched threads list.
* **watch_email** - put on watched threads list and send e-mail when somebody replies.


### subscribe_start

Default value for automatic subscription to started threads prefference for new user accounts. Allows for same values as `subscribe_reply`.


### thread_title_length_max

Maximal allowed thread title length.


### thread_title_length_min

Minimal allowed thread title length.


### username_length_max

Maximal allowed username length.


### username_length_min

Minimal allowed username length.


# Django Settings Reference

Django defines plenty of configuration options that control behaviour of different features that Misago relies on.

Those are documented and available in [Django documentation](https://docs.djangoproject.com/en/1.6/ref/settings/).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 2021-02-09

### Changed
- [PG-1313](https://bighealth.atlassian.net/browse/PG-1313) Remove the online status ([#14](https://github.com/sleepio/community-app-web/pull/14))

## 2021-02-06

### Changed
- [PG-1428](https://bighealth.atlassian.net/browse/PG-1428) Remove 'decline' button from posting policy ([#13](https://github.com/sleepio/community-app-web/pull/13))


devproject/settings.py
# pylint: disable=line-too-long
"""
Django settings for dev project.

Generated by 'django-admin startproject' using Django 1.11.15.

For more information on this file, see
https://docs.djangoproject.com/en/1.11/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.11/ref/settings/
"""

import os
from jslog4kube import LOGGING
from logging.config import dictConfig

from misago import load_plugin_list_if_exists


# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.abspath(__file__))


# Define placeholder gettext function
# This function will mark strings in settings visible to makemessages
# without need for Django's i18n features be initialized first.
_ = lambda s: s


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.11/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "1znyfpwp*_#!r0#l248lht*6)_0b+504n*2-8cxf(2u)fhi0f^"

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = os.getenv('DEBUG', True) in ('True', 'true', 'yes')


# A list of strings representing the host/domain names that this Django site can serve.
# If you are unsure, just enter here your domain name, eg. ['mysite.com', 'www.mysite.com']

ALLOWED_HOSTS = ["localhost", "127.0.0.1", "0.0.0.0", "*"]


# Database
# https://docs.djangoproject.com/en/1.11/ref/settings/#databases

DATABASES = {
    "default": {
        # Misago requires PostgreSQL to run
        "ENGINE": "django.db.backends.postgresql",
        "NAME": os.environ.get("POSTGRES_DB"),
        "USER": os.environ.get("POSTGRES_USER"),
        "PASSWORD": os.environ.get("POSTGRES_PASSWORD"),
        "HOST": os.environ.get("POSTGRES_HOST"),
        "PORT": 5432,
    }
}


# Caching
# https://docs.djangoproject.com/en/1.11/topics/cache/#setting-up-the-cache

CACHES = {
    "default": {
        # Misago doesn't run well with LocMemCache in production environments
        "BACKEND": "django.core.cache.backends.locmem.LocMemCache"
    }
}


# Password validation
# https://docs.djangoproject.com/en/1.11/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
        "OPTIONS": {"user_attributes": ["username", "email"]},
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
        "OPTIONS": {"min_length": 7},
    },
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},
    {"NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"},
]


# Internationalization
# https://docs.djangoproject.com/en/1.11/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.11/howto/static-files/

STATIC_URL = "/static/"


# User uploads (Avatars, Attachments, files uploaded in other Django apps, ect.)
# https://docs.djangoproject.com/en/1.11/howto/static-files/

MEDIA_URL = "/media/"


# The absolute path to the directory where collectstatic will collect static files for deployment.
# https://docs.djangoproject.com/en/1.11/ref/settings/#static-root

STATIC_ROOT = os.path.join(BASE_DIR, "static")


# Absolute filesystem path to the directory that will hold user-uploaded files.
# https://docs.djangoproject.com/en/1.11/ref/settings/#media-root

MEDIA_ROOT = os.path.join(BASE_DIR, "media")


# This setting defines the additional locations the staticfiles app will traverse if the FileSystemFinder finder
# is enabled, e.g. if you use the collectstatic or findstatic management command or use the static file serving view.
# https://docs.djangoproject.com/en/1.10/ref/settings/#staticfiles-dirs

STATICFILES_DIRS = []


# Email configuration
# https://docs.djangoproject.com/en/1.11/ref/settings/#email-backend

EMAIL_HOST = "localhost"
EMAIL_PORT = 25


# If either of these settings is empty, Django won't attempt authentication.

EMAIL_HOST_USER = ""
EMAIL_HOST_PASSWORD = ""


# Default email address to use for various automated correspondence from the site manager(s).

DEFAULT_FROM_EMAIL = "Forums <%s>" % EMAIL_HOST_USER


# Application definition

AUTH_USER_MODEL = "misago_users.User"

AUTHENTICATION_BACKENDS = ["misago.users.authbackends.MisagoBackend"]

CSRF_FAILURE_VIEW = "misago.core.errorpages.csrf_failure"

PLUGINS_LIST_PATH = os.path.join(os.path.dirname(BASE_DIR), "plugins.txt")

INSTALLED_PLUGINS = load_plugin_list_if_exists(PLUGINS_LIST_PATH) or []

INSTALLED_APPS = INSTALLED_PLUGINS + [
    # Misago overrides for Django core feature
    "misago",
    "misago.users",
    # Django apps
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.postgres",
    "django.contrib.humanize",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    # 3rd party apps used by Misago
    "ariadne.contrib.django",
    "celery",
    "django_extensions",
    "mptt",
    "rest_framework",
    "social_django",
    # Misago apps
    "misago.admin",
    "misago.acl",
    "misago.analytics",
    "misago.cache",
    "misago.core",
    "misago.conf",
    "misago.icons",
    "misago.themes",
    "misago.markup",
    "misago.legal",
    "misago.categories",
    "misago.threads",
    "misago.readtracker",
    "misago.search",
    "misago.socialauth",
    "misago.graphql",
    "misago.faker",
    "misago.menus",
    "misago.sso",
    "misago.plugins",
]

if DEBUG:
    INSTALLED_APPS += [
     "debug_toolbar",   
    ]

INTERNAL_IPS = ["127.0.0.1"]

LOGIN_REDIRECT_URL = "misago:index"

LOGIN_URL = "misago:login"

LOGOUT_URL = "misago:logout"

MIDDLEWARE = [
    "misago.users.middleware.RealIPMiddleware",
    "misago.core.middleware.FrontendContextMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "misago.cache.middleware.cache_versions_middleware",
    "misago.conf.middleware.dynamic_settings_middleware",
    "misago.socialauth.middleware.socialauth_providers_middleware",
    "misago.users.middleware.UserMiddleware",
    "misago.acl.middleware.user_acl_middleware",
    "misago.core.middleware.ExceptionHandlerMiddleware",
    "misago.users.middleware.OnlineTrackerMiddleware",
    "misago.admin.middleware.AdminAuthMiddleware",
    "misago.threads.middleware.UnreadThreadsCountMiddleware",
]

if DEBUG:
    MIDDLEWARE = [
        "debug_toolbar.middleware.DebugToolbarMiddleware",
    ] + MIDDLEWARE

ROOT_URLCONF = "devproject.urls"

SOCIAL_AUTH_STRATEGY = "misago.socialauth.strategy.MisagoStrategy"

SOCIAL_AUTH_PIPELINE = (
    # Steps required by social pipeline to work - don't delete those!
    "social_core.pipeline.social_auth.social_details",
    "social_core.pipeline.social_auth.social_uid",
    "social_core.pipeline.social_auth.social_user",
    # If enabled in admin panel, lets your users to associate their old forum account
    # with social one, if both have same e-mail address.
    "misago.socialauth.pipeline.associate_by_email",
    # Those steps make sure banned users may not join your site or use banned name or email.
    "misago.socialauth.pipeline.validate_ip_not_banned",
    "misago.socialauth.pipeline.validate_user_not_banned",
    # Reads user data received from social site and tries to create valid and available username
    # Required if you want automatic account creation to work. Otherwise optional.
    "misago.socialauth.pipeline.get_username",
    # Uncomment next line to enable automatic account creation if data from social site is valid
    # and get_username found valid name for new user account:
    # 'misago.socialauth.pipeline.create_user',
    # This step asks user to complete simple, pre filled registration form containing username,
    # email, legal note if you remove it without adding custom one, users will have no fallback
    # for joining your site using their social account.
    "misago.socialauth.pipeline.create_user_with_form",
    # Steps finalizing social authentication flow - don't delete those!
    "social_core.pipeline.social_auth.associate_user",
    "social_core.pipeline.social_auth.load_extra_data",
    "misago.socialauth.pipeline.require_activation",
)

SOCIAL_AUTH_POSTGRES_JSONFIELD = True

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.i18n",
                "django.template.context_processors.media",
                "django.template.context_processors.request",
                "django.template.context_processors.static",
                "django.template.context_processors.tz",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
                "misago.acl.context_processors.user_acl",
                "misago.conf.context_processors.conf",
                "misago.conf.context_processors.og_image",
                "misago.core.context_processors.misago_version",
                "misago.core.context_processors.request_path",
                "misago.core.context_processors.momentjs_locale",
                "misago.icons.context_processors.icons",
                "misago.search.context_processors.search_providers",
                "misago.themes.context_processors.theme",
                "misago.legal.context_processors.legal_links",
                "misago.menus.context_processors.menus",
                "misago.users.context_processors.user_links",
                "misago.core.context_processors.hooks",
                # Data preloaders
                "misago.conf.context_processors.preload_settings_json",
                "misago.core.context_processors.current_link",
                "misago.markup.context_processors.preload_api_url",
                "misago.threads.context_processors.preload_threads_urls",
                "misago.users.context_processors.preload_user_json",
                "misago.socialauth.context_processors.preload_socialauth_json",
                # Note: keep frontend_context processor last for previous processors
                # to be able to expose data UI app via request.frontend_context
                "misago.core.context_processors.frontend_context",
            ]
        },
    }
]

WSGI_APPLICATION = "devproject.wsgi.application"


# Django Debug Toolbar
# http://django-debug-toolbar.readthedocs.io/en/stable/configuration.html

if DEBUG:
    DEBUG_TOOLBAR_PANELS = [
        "debug_toolbar.panels.versions.VersionsPanel",
        "debug_toolbar.panels.timer.TimerPanel",
        "debug_toolbar.panels.settings.SettingsPanel",
        "debug_toolbar.panels.headers.HeadersPanel",
        "debug_toolbar.panels.request.RequestPanel",
        "debug_toolbar.panels.sql.SQLPanel",
        "misago.acl.panels.MisagoACLPanel",
        "debug_toolbar.panels.staticfiles.StaticFilesPanel",
        "debug_toolbar.panels.templates.TemplatesPanel",
        "debug_toolbar.panels.cache.CachePanel",
        "debug_toolbar.panels.signals.SignalsPanel",
        "debug_toolbar.panels.logging.LoggingPanel",
    ]


# Django Rest Framework
# http://www.django-rest-framework.org/api-guide/settings/

REST_FRAMEWORK = {
    "DEFAULT_PERMISSION_CLASSES": [
        "misago.core.rest_permissions.IsAuthenticatedOrReadOnly"
    ],
    "DEFAULT_RENDERER_CLASSES": ["rest_framework.renderers.JSONRenderer"],
    "EXCEPTION_HANDLER": "misago.core.exceptionhandler.handle_api_exception",
    "UNAUTHENTICATED_USER": "misago.users.models.AnonymousUser",
    "URL_FORMAT_OVERRIDE": None,
}


# Celery - Distributed Task Queue
# http://docs.celeryproject.org/en/latest/userguide/configuration.html

# Configure Celery to use Redis as message broker.

CELERY_BROKER_URL = "redis://redis:6379/0"

# Celery workers may leak the memory, eventually depriving the instance of resources.
# This setting forces celery to stop worker, clean after it and create new one
# after worker has processed 10 tasks.

CELERY_WORKER_MAX_TASKS_PER_CHILD = 10


# Misago specific settings
# https://misago.readthedocs.io/en/latest/developers/settings.html

# On dev instance, generate only three sizes of avatars instead of default 6 sizes.

MISAGO_AVATARS_SIZES = [400, 200, 100]


# PostgreSQL text search configuration to use in searches
# Defaults to "simple", for list of installed configurations run "\dF" in "psql".
# Standard configs as of PostgreSQL 9.5 are: dutch, english, finnish, french,
# german, hungarian, italian, norwegian, portuguese, romanian, russian, simple,
# spanish, swedish and turkish
# Example on adding custom language can be found here: https://github.com/lemonskyjwt/plpstgrssearch

MISAGO_SEARCH_CONFIG = "simple"


# Path to the directory that Misago should use to prepare user data downloads.
# Should not be accessible from internet.

MISAGO_USER_DATA_DOWNLOADS_WORKING_DIR = os.path.join(BASE_DIR, "userdata")


# Path to directory containing avatar galleries
# Those galleries can be loaded by running loadavatargallery command

MISAGO_AVATAR_GALLERY = os.path.join(BASE_DIR, "avatargallery")


# Profile fields

MISAGO_PROFILE_FIELDS = [
    {
        "name": _("Personal"),
        "fields": [
            "misago.users.profilefields.default.RealNameField",
            "misago.users.profilefields.default.GenderField",
            "misago.users.profilefields.default.BioField",
            "misago.users.profilefields.default.LocationField",
        ],
    },
    {
        "name": _("Contact"),
        "fields": [
            "misago.users.profilefields.default.TwitterHandleField",
            "misago.users.profilefields.default.SkypeIdField",
            "misago.users.profilefields.default.WebsiteField",
        ],
    },
    {
        "name": _("IP address"),
        "fields": ["misago.users.profilefields.default.JoinIpField"],
    },
]


# Set dev instance to send e-mails to console

# EMAIL_BACKEND = "django.core.mail.backends.console.EmailBackend"
EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
EMAIL_HOST = 'mailhog'
EMAIL_PORT = 1025

# Display debug toolbar if IN_MISAGO_DOCKER enviroment var is set to "1"
if DEBUG:
    DEBUG_TOOLBAR_CONFIG = {
        "SHOW_TOOLBAR_CALLBACK": "misago.conf.debugtoolbar.enable_debug_toolbar"
    }


LOGGERS = {
    "pagesnap": {
        "handlers": ["json-stdout"],
        "formatters": ["json"],
        "propagate": True,
        "level": "DEBUG",
    }
}

LOGGING["loggers"].update(LOGGERS)

dictConfig(LOGGING)